service/src/main/java/uk/gov/hmcts/reform/fpl/service/cmo/UploadCMOService.java

```diff
@@ -0,0 +1,164 @@
+    private static final String CMO_JUDGE_INFO_FIELD = "cmoJudgeInfo";
+    private static final String PAST_HEARING_LIST_FIELD = "pastHearingList";
+
+    private static final String SINGLE = "SINGLE";
+    private static final String MULTI = "MULTI";
+    private static final String NONE = "NONE";
+    private static final String NUM_HEARINGS_FIELD = "numHearings";
+
+    private final ObjectMapper mapper;
+
+    public Map<String, Object> getInitialPageData(List<Element<HearingBooking>> hearings,
+                                                  List<Element<CaseManagementOrder>> unsealedOrders) {
+
+        List<Element<HearingBooking>> hearingsWithoutCMOs = getHearingsWithoutCMO(hearings, unsealedOrders);
+
+        Map<String, Object> data = new HashMap<>();
+        String textAreaKey = null;
+        String numHearings;
+        String showTextAreaKey = null;
+
+        switch (hearingsWithoutCMOs.size()) {
+            case 0:
+                numHearings = NONE;
+                break;
+            case 1:
+                numHearings = SINGLE;
+                textAreaKey = "singleHearingsWithCMOs";
+                showTextAreaKey = "showHearingsSingleTextArea";
+                data.putAll(getJudgeAndHearingDetailsSingle(hearingsWithoutCMOs.get(0).getId(), hearingsWithoutCMOs));
+                break;
+            default:
+                numHearings = MULTI;
+                textAreaKey = "multiHearingsWithCMOs";
+                showTextAreaKey = "showHearingsMultiTextArea";
+                data.put(PAST_HEARING_LIST_FIELD, buildDynamicList(hearingsWithoutCMOs));
+        }
+
+        String textAreaContent = buildHearingsWithCMOsText(unsealedOrders, hearings);
+
+        if (textAreaContent.length() != 0 && textAreaKey != null) {
+            data.put(textAreaKey, textAreaContent);
+            data.put(showTextAreaKey, "YES");
+        }
+
+        data.put(NUM_HEARINGS_FIELD, numHearings);
+
+        return data;
+    }
+
+    public List<Element<HearingBooking>> getHearingsWithoutCMO(List<Element<HearingBooking>> hearings,
+                                                               List<Element<CaseManagementOrder>> unsealedOrders) {
+        return hearings.stream()
+            .filter(hearing -> associatedToReturnedCMO(hearing, unsealedOrders)
+                || !hearing.getValue().hasCMOAssociation())
+            .collect(toList());
+    }
+
+    public Map<String, Object> getJudgeAndHearingDetails(UUID selectedHearing,
+                                                         List<Element<HearingBooking>> hearings) {
+        HearingBooking selected = getSelectedHearing(selectedHearing, hearings);
+        return Map.of(
+            CMO_JUDGE_INFO_FIELD, formatJudgeTitleAndName(selected.getJudgeAndLegalAdvisor()),
+            CMO_HEARING_INFO_FIELD, selected.toLabel(DATE)
+        );
+    }
+
+    public UUID mapToHearing(UUID selectedHearing, List<Element<HearingBooking>> hearings,
```

Comment:
I don't like this method seems functional (mapToHreaing), but performs set, it's side-effect which I wouldn't expect here. Is the side effect required? If yes could you consider renaming and make it explicit by name what's really happening?

Reply:
refactored and renamed to `updateHearingWithCmoId`